src/compiler/checker.ts

```diff
@@ -10305,11 +10400,17 @@ namespace ts {
 
         function getTypeAtPosition(signature: Signature, pos: number): Type {
             return signature.hasRestParameter ?
-                pos < signature.parameters.length - 1 ? getTypeOfSymbol(signature.parameters[pos]) : getRestTypeOfSignature(signature) :
-                pos < signature.parameters.length ? getTypeOfSymbol(signature.parameters[pos]) : anyType;
+                getParameterTypeAtIndex(signature, pos, signature.parameters.length - 1) :
+                getParameterTypeAtIndex(signature, pos, signature.parameters.length, anyType);
         }
 
         function assignContextualParameterTypes(signature: Signature, context: Signature, mapper: TypeMapper) {
+            if (context.thisType) {
+                if (signature.declaration.kind !== SyntaxKind.ArrowFunction) {
+                    // do not contextually type thisType for ArrowFunction. 
+                    signature.thisType = context.thisType;
+                }
+            }

```

Comment:
I think the problem is that your implementation doesn't work in "pull mode". Try pasting that example into VS Code and hover over `this`. You'll see type `any`. This is because something needs to have "pushed" a this type into `m`'s signature before you can pull it out in a check of `this`.